### PR TITLE
Detector Copilot Feature

### PR DESCRIPTION
**Pull request by ShekharGupta1988**
This Pull Request introduces the Detector Copilot feature with both backend and frontend changes. 
There are quite a few changes but to summarize, below are the major updates:

### Backend Updates:
 
- `Controllers\OpenAIController.cs` : Added HTTP GET method for ChatCompletion API
- `Hubs\OpenAIChatCompletionHub.cs`: Introduced a SignalR message hub for streaming chat messages.
- `Services\OpenAIService\OpenAIService.cs`: Implemented support for GPT-4 APIs using the OpenAI SDK, enabling both streaming and non-streaming capabilities.
- `OpenAIChatTemplates\` : Created a folder to store prompt templates.
- Integrated SignalR support into the Applens Backend project.

### Frontend Updates:
 

`openai-chat component`: Updated the chat component that supports GPT-3, GPT-4, WebSocket, and REST API protocols, along with the ability to cancel in-flight messages.
`applens-openai-chat.service`: Incorporated support for GPT-4 HTTP and SignalR connections, as well as event listeners.
`detector-copilot component`: Established a copilot component to manage detector code and non-code related messages while interacting with onboarding modules.
`detector-copilot service` : Designed a copilot service to maintain copilot state and facilitate communication between the copilot and other components.

### Additional Updates:
- Implemented a collapsible side menu for Applens.
- Some code cleanup in existing components to improve maintainability and readability.

### CODE REVIEW

File: AngularApp/projects/applens/src/app/modules/dashboard/detector-copilot/detector-copilot.component.ts [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+      Detector Copilot (Preview)
+      <img class='copilot-header-img-secondary' src='/assets/img/rocket.png' alt=''>
+      <img class='copilot-header-img-secondary' src='/assets/img/rocket.png' alt=''">
+    </h1>`;
+  }
+
+  private prepareCodeUpdateMessages = () => {
+
+    this.http.get<any>(this.configFile).subscribe(res => {
+      this.codeProgressMessages = res && res.codeProgressMessages && res.codeProgressMessages.length > 0 ?
```
Comment (osshad) [inline AngularApp/projects/applens/src/app/modules/dashboard/detector-copilot/detector-copilot.component.ts:350]: This is just a suggestion (not a required change). Going forward, I think we can use optional chaining to simplify these conditional checks.

Comment (ShekharGupta1988) [reply AngularApp/projects/applens/src/app/modules/dashboard/detector-copilot/detector-copilot.component.ts:350]: good suggestion. I'll fix it here and other places. Makes the code smaller :)

Comment (ShekharGupta1988) [reply AngularApp/projects/applens/src/app/modules/dashboard/detector-copilot/detector-copilot.component.ts:350]: good suggestion. I'll fix it here and other places. Makes the code smaller :)

File: AngularApp/package-lock.json
Comment (SharmaAjay19) [inline AngularApp/package-lock.json:1]: I see this file in .gitignore but still shows up here.

**Review by SharmaAjay19 — APPROVED** [review]

**Review by osshad — APPROVED** [review]
Thanks for making those changes. LGTM 🚀.  Great job, Shekhar!! 🔥

**Comment by hforeste** [comment]
Excited for this project, curious if you ran Applens in National Cloud mode to see if this works fine or doesn't compromise the existing experience.

**Comment by ShekharGupta1988** [comment]
> Excited for this project, curious if you ran Applens in National Cloud mode to see if this works fine or doesn't compromise the existing experience.

yup, I ran the project with ncloud setting and existing detectors were working fine.

**Pull request merged**